libkbfs/dirty_file.go

```diff
@@ -0,0 +1,159 @@
+	// writes.
+	fileBlockStates map[BlockPointer]dirtyBlockState
+}
+
+func newDirtyFile() *dirtyFile {
+	return &dirtyFile{
+		fileBlockStates: make(map[BlockPointer]dirtyBlockState),
+	}
+}
+
+func (df *dirtyFile) blockNeedsCopy(ptr BlockPointer) bool {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	return df.fileBlockStates[ptr].copy == blockNeedsCopy
+}
+
+// setBlockDirty transitions a block to a dirty state, and returns
+// whether or not the block needs to be put in the dirty cache
+// (because it isn't yet), and whether or not the block is currently
+// part of a sync in progress.
+func (df *dirtyFile) setBlockDirty(ptr BlockPointer) (
+	needsCaching bool, isSyncing bool) {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+
+	state := df.fileBlockStates[ptr]
+	needsCaching = state.copy == blockNeedsCopy
+	state.copy = blockAlreadyCopied
+	isSyncing = state.sync != blockNotSyncing
+	df.fileBlockStates[ptr] = state
+	return needsCaching, isSyncing
+}
+
+func (df *dirtyFile) isBlockSyncing(ptr BlockPointer) bool {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	return df.fileBlockStates[ptr].sync == blockSyncing
+}
+
+func (df *dirtyFile) isBlockDirty(ptr BlockPointer) bool {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	return df.fileBlockStates[ptr].copy == blockAlreadyCopied
+}
+
+func (df *dirtyFile) setBlockSyncing(ptr BlockPointer) error {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	state := df.fileBlockStates[ptr]
+	if state.copy == blockNeedsCopy {
+		return fmt.Errorf("Trying to sync a block that isn't dirty: %v", ptr)
+	}
+	state.copy = blockNeedsCopy
+	state.sync = blockSyncing
+	df.fileBlockStates[ptr] = state
+	return nil
+}
+
+func (df *dirtyFile) resetSyncingBlocksToDirty() {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	// Reset all syncing blocks to just be dirty again
+	for ptr, state := range df.fileBlockStates {
+		if state.sync != blockNotSyncing {
+			state.copy = blockAlreadyCopied
+			state.sync = blockNotSyncing
+			df.fileBlockStates[ptr] = state
+		}
+	}
+}
+
+func (df *dirtyFile) setBlockSyncedLocked(ptr BlockPointer) error {
+	state := df.fileBlockStates[ptr]
+	if state.copy == blockAlreadyCopied && state.sync == blockNotSyncing {
+		// We've likely already had an resetSyncingBlocksToDirty call; ignore.
+		return nil
+	}
+
+	if state.sync != blockSyncing {
+		return fmt.Errorf("Trying to finish a block sync that wasn't in "+
+			"progress: %v (%d)", ptr, df.fileBlockStates[ptr])
+	}
+	state.sync = blockSynced
+	df.fileBlockStates[ptr] = state
+	// TODO: Eventually we'll need to free up space in the buffer
+	// taken up by these sync'd blocks, so new writes can proceed.
+	return nil
+}
+
+func (df *dirtyFile) setBlockSynced(ptr BlockPointer) error {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	return df.setBlockSyncedLocked(ptr)
+}
+
+func (df *dirtyFile) finishSync() {
+	// Mark any remaining blocks as finished syncing.  For example,
+	// top-level indirect blocks needs this because they are added to
```

Comment:
This is the only case, right? If so, might be worth making this function verify that condition, i.e. have `finishSync` take in the top `BlockPointer` and verify that it's the only block left in `blockSyncing`, and then set it to synced.